lib/ide/file-manager/mock-server/src/main.rs

```diff
@@ -0,0 +1,228 @@
+            }
+            Err(e) => {
+                messages::Result::new_error_simple(FAILED_CALL_ERROR_CODE,e.to_string())
+            }
+        }
+    }
+
+    /// Takes JSON-encoded message and returns the response.
+    fn handle_message(&self, request_text:String) -> serde_json::Value {
+        let request = serde_json::from_str::<SomeRequest>(&request_text);
+        match request {
+            Ok(request) => {
+                let id       = request.id;
+                let result   = self.handle_request(request);
+                let response = messages::Response{id,result};
+                let reply    = messages::Message::new(response);
+                serde_json::to_value(reply).unwrap()
+            },
+            Err(_e) => {
+                // TODO: [mwu] here should go proper error handling that builds
+                //       an error that has `id` set to `null`
+                default()
+            }
+        }
+    }
+
+    /// Reads text messages and replies to them.
+    /// If the connection is lost, silently returns.
+    fn run(&mut self) {
+        while let Ok(request_msg) = self.socket.read_message() {
+            if request_msg.is_text() {
+                let request_text = request_msg.to_string();
+                let reply_json   = self.handle_message(request_text);
+                let reply_text   = serde_json::from_value::<String>(reply_json).unwrap();
+                let reply_msg    = Message::text(reply_text);
+                if self.socket.write_message(reply_msg).is_err() {
+                    return;
+                }
+            }
+            // ignore non-text messages.
+        }
+    }
+}
+
+
+
+// ==========
+// == Main ==
+// ==========
+
+fn main() {
+    let port    = utils::env::parse_var_or(PORT_VAR, DEFAULT_PORT);
+    let address = format!("127.0.0.1:{}", port);
+    let server  = TcpListener::bind(&address).unwrap();
+    println!("Listening on {}", address);
+    for stream in server.incoming() {
+        spawn(move || {
+            println!("Got a new connection");
+            let callback = |_req: &Request, response: Response| {
+                Ok(response)
+            };
+            let socket = accept_hdr(stream.unwrap(), callback).unwrap();
+            let mut server = Handler { socket };
+            server.run();
+        });
+    }
+}
+
+
+
+// =============
+// == IsCall  ==
+// =============
+
+/// All methods supported by this server should implement the trait.
+pub trait IsCall {
+    /// Type of value returned on a successful call.
+    type Result: Serialize;
+
+    /// Perform the "actual" work for this call.
+    fn realize(&self) -> CallResult<Self::Result>;
+}
+
+
+
+// ===========
+// == Call  ==
+// ===========
+
+/// Possible call requests. Members should wrap structs implementing `IsCall`.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+#[serde(tag = "method", content = "params", rename_all = "camelCase")]
```

Comment:
We dont put spaces neither for derive (as it is like function call), nor around `=` in derive mechanisms